test/extended/dr/machine_recover.go

```diff
@@ -0,0 +1,276 @@
+				o.Expect(len(healthyEtcds)).To(o.BeNumerically("==", len(masters)))
+
+			})
+	},
+	)
+})
+
+func getHealthyEtcds(certsDir string, masters []*corev1.Node) []*etcdserverpb.Member {
+	endpoints := getEtcdEndpoints(masters)
+	etcdClient, err := getEtcdClient(certsDir, endpoints)
+	o.Expect(err).ToNot(o.HaveOccurred())
+	memberListResp, err := etcdClient.MemberList(context.Background())
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	healthyMembers := []*etcdserverpb.Member{}
+
+	for _, m := range memberListResp.Members {
+		_, err := etcdClient.Status(context.Background(), m.ClientURLs[0])
+		if err != nil {
+			healthyMembers = append(healthyMembers, m)
+		}
+	}
+	return healthyMembers
+}
+
+func getEtcdClient(certsDir string, endpoints []string) (*clientv3.Client, error) {
+	dialOptions := []grpc.DialOption{
+		grpc.WithBlock(), // block until the underlying connection is up
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      path.Join(certsDir, "tls.crt"),
+		KeyFile:       path.Join(certsDir, "tls.key"),
+		TrustedCAFile: path.Join(certsDir, "ca-bundle.crt"),
+	}
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	cfg := &clientv3.Config{
+		DialOptions: dialOptions,
+		Endpoints:   endpoints,
+		DialTimeout: 15 * time.Second,
+		TLS:         tlsConfig,
+	}
+
+	cli, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cli, err
+}
+
+func dumpEtcdCertsOnDisk(oc *exutil.CLI, dir string) {
+	err := os.MkdirAll(dir, os.ModePerm)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	etcdCA, err := oc.AdminKubeClient().CoreV1().ConfigMaps("openshift-config").Get("etcd-ca-bundle", metav1.GetOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+	caData, ok := etcdCA.Data["ca-bundle.crt"]
+	if !ok {
+		o.Expect(fmt.Errorf("etcd CA data  missing in configmap openshift-config/etcd-ca-bundle")).ToNot(o.HaveOccurred())
+	}
+
+	etcdClientCerts, err := oc.AdminKubeClient().CoreV1().Secrets("openshift-config").Get("etcd-client", metav1.GetOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+	clientCert, ok := etcdClientCerts.Data["tls.crt"]
+	if !ok {
+		o.Expect(fmt.Errorf("etcd client Certificate data  missing in secret openshift-config/etcd-client")).ToNot(o.HaveOccurred())
+	}
+	clientKey, ok := etcdClientCerts.Data["tls.key"]
+	if !ok {
+		o.Expect(fmt.Errorf("etcd client Private Key data  missing in secret openshift-config/etcd-client")).ToNot(o.HaveOccurred())
+	}
+
+	err = ioutil.WriteFile(path.Join(dir, "ca-bundle.crt"), []byte(caData), 0600)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	err = ioutil.WriteFile(path.Join(dir, "tls.crt"), []byte(clientCert), 0600)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	err = ioutil.WriteFile(path.Join(dir, "tls.key"), []byte(clientKey), 0600)
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+func getEtcdEndpoints(masters []*corev1.Node) []string {
+	endpoints := []string{}
+	for _, m := range masters {
+		for _, addr := range m.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP {
```

Comment:
From where tests are run do we have access to the node network? If not we would have to ssh into the master and run etcd health check

Reply:
I see `expectSSH` function using node's `ExternalIP` to ssh and if not available, using the `InternalIP`. Can we use the same strategy for a non-ssh port i. e 2379?

Reply:
we don't dropped